executor/src/witgen/machines/block_machine.rs

```diff
@@ -591,18 +644,125 @@ impl<'a, T: FieldElement> WitnessColumnEvaluator<T> for WitnessData<'a, T> {
+    current_round_count: usize,
+}
+
+impl DefaultSequenceIteratorState {
+    fn new() -> Self {
+        DefaultSequenceIteratorState {
+            is_first: true,
+            progress_in_current_round: false,
+            cur_row_delta_index: 0,
+            cur_identity_index: 0,
+            current_round_count: 0,
+        }
+    }
+
+    /// Update the state of the iterator.
+    /// If we're not at the last identity in the current row, just moves to the next.
+    /// Otherwise, starts with identity 0 and moves to the next row if no progress was made.
+    fn update(&mut self, is_last_identity: bool, progress_in_last_step: bool) {
+        if !self.is_first {
+            self.progress_in_current_round |= progress_in_last_step;
+            if is_last_identity {
+                // Start next round
+                if !self.progress_in_current_round || self.current_round_count > 20 {
+                    // Move to next row delta, starting with identity 0.
+                    if self.current_round_count > 20 {
+                        log::warn!("In witness generation for block machine, we have been stuck in the same row for 20 rounds. \
+                                    This is likely a bug in the witness generation algorithm.");
+                    }
+
+                    self.cur_row_delta_index += 1;
+                    self.current_round_count = 0;
+                } else {
+                    // Stay and current row delta, starting with identity 0.
+                    self.current_round_count += 1;
+                }
+                self.cur_identity_index = 0;
+                self.progress_in_current_round = false;
+            } else {
+                // Stay at row delta, move to next identity.
+                self.cur_identity_index += 1;
+            }
+        }
+        self.is_first = false;
+    }
+}
+
+/// Goes through all rows of the block machine (plus the ones before and after)
+/// forward, backward, and forward again.
+/// In each row, iterates over all identities until no further progress is made.
+struct DefaultSequenceIterator {
+    block_size: usize,
+    identities_count: usize,
+    row_deltas: Vec<i64>,
+    state: DefaultSequenceIteratorState,
+}
+
+impl DefaultSequenceIterator {
+    fn new(block_size: usize, identities_count: usize) -> Self {
+        let block_size_i64 = block_size as i64;
+        DefaultSequenceIterator {
+            block_size,
+            identities_count,
+            row_deltas: (-1..=block_size_i64)
+                .chain((-1..block_size_i64).rev())
+                .chain(0..=block_size_i64)
+                .collect(),
+            state: DefaultSequenceIteratorState::new(),
+        }
+    }
+
+    fn next(&mut self, progress_in_last_step: bool) -> Option<SequenceStep> {
+        let row_delta = self.row_deltas[self.state.cur_row_delta_index];
+        let is_last_identity = if row_delta + 1 == self.block_size as i64 {
```

Comment:
I find it very confusing that you compute a parameter that you pass to the state from information internal to the state...